_posts/2017-07-06-hitchhikers-guide-to-cp.md

```diff
@@ -0,0 +1,308 @@
+
++   **Memory limit :** Given in *MB*. Maximum amount of memory your code should use.
+
++   **Input Size :** Maximum possible input size, ie, input bigger than this won't be there in test cases.
+
+Then, there will be *Sample Input* and *Sample Output* to give you an idea of I/O formats and to give you a better understanding of the problem.
+
+Lastly, there will be *Explanation*(might be absent) which explains how sample output is calculated.
+
+## Solving the problem:
+
+After reading and understanding the problem, a solution is to be coded for it. A given problem can have many possible solutions but you have to find that solution which comes under the constraints. In order to see if your code fulfills the time and memory limits, you need to do an *Order Analysis* of your solution.
+
+**Order Analysis :** 
+Order of a solution is a measure of it's runtime or number of operations as a function of input size. Let's take an example, Suppose we have to find the maximum among N positive numbers.
+
+```
+int ans = -1,x;
+for(int i=0;i<N;i++) {
+    scanf(“%d”,&x);
+    ans = max(ans, x);
+}
+```
+
+N times checking i < N.
+
+N times checking max of ans and x.
+
+N times i++.
+
+N times scanf.
+
+Hence, total number of operations is $$4 * N$$. We only care about dependency on N. Therefore, order is $$O(N)$$.
+
+[CLRS](https://www.google.co.in/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwju8crC0e_UAhWHN48KHTAVCdoQFggjMAA&url=https%3A%2F%2Fweb.njit.edu%2F~wl256%2Fdownload%2Fcs610%2FIntroduction-to-algorithm-3rdEdition.pdf&usg=AFQjCNGJTrGFQdXhlY7ItJk2pCGX9V5S9Q) is having detailed theoretical explanation on Order Analysis.
+Also, for most of the online judges(SPOJ, Codechef etc), number of operations in 1 sec is approximately $$10^8$$.
+In some problems, there are T test cases in one input file. If a solution is having order of $$O(N^3)$$, then total order of the whole input will be $$O(T * N^3)$$.
+
+For Example, if T < 100 and N < $$10^5$$, and order per test case is $$O(N)$$, then number of operations will be at max $$10^5$$ per test case and $$10^7$$ per input file.
+
+## Verdict:
+
+Once you submit your solution, the judge will run a few test cases on your solution and will give you a verdict based on the output of your solution.
+
+**There will be one of following verdicts :**
++   **AC**(Also shown by a green tick): Your solution is accepted.
+
++   **WA**(Also shown by a red cross): Your code gives wrong output. Check for correctness of your approach.
+
++   **TLE**(Also shown by a clock): Your code is taking more time than the given limit. You need to code a faster solution.
+
++   **Run time error**: This could occur due to-
+
+    +   Segmentation fault which could be due to declaring array of size smaller than required or accessing negative indices.
+
+    +   Dividing by zero.
+
+    +   Declaring arrays of size more than $$10^8$$.
```

Comment:
Could we mention the fact that declaring arrays inside main puts them in the stack and outside puts them in the heap, leading to segfaults for smaller sizes if declared inside main?